## Setup

In [2]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import sklearn as sk
from sklearn.linear_model import LinearRegression

In [3]:
df = pd.read_csv("Lab1/Advertising.csv")

In [4]:
X = df[['TV', 'radio', 'newspaper']] # extract relevant features
print(X) # print dataframe
X = np.c_[X, np.ones(len(X.index))] # convert to numpy matrix + the last columns being 1s to account for 'intercept'

        TV  radio  newspaper
0    230.1   37.8       69.2
1     44.5   39.3       45.1
2     17.2   45.9       69.3
3    151.5   41.3       58.5
4    180.8   10.8       58.4
..     ...    ...        ...
195   38.2    3.7       13.8
196   94.2    4.9        8.1
197  177.0    9.3        6.4
198  283.6   42.0       66.2
199  232.1    8.6        8.7

[200 rows x 3 columns]


In [5]:
Y = df[['sales']] # extract sales
print(Y) # print dataframe
Y = np.c_[Y] # convert to numpy matrix

     sales
0     22.1
1     10.4
2      9.3
3     18.5
4     12.9
..     ...
195    7.6
196    9.7
197   12.8
198   25.5
199   13.4

[200 rows x 1 columns]


## Problem 1

#### 1. Solve the Normal Equation to find the model parameters θ_i (for 0 ≤ i ≤ N)

sale = θ_0 + θ_1(TV) + θ_2(radio) + θ_3(newspaper) 

I will first rewrite the equation: Y_i = (X_i)^T * B + e_i

Where:
- B, "beta" is a matrix of the coefficents 
- X_i, is a feature vector of the ith observation
- Y_i, the results vector

**Normal equation** (we want to solve for B): 
- (X^T * X) * B = X^T * Y
- B = (X^T * X)^-1 * (X^T * Y)

In [11]:
# np.linlag.inv Computes the (multiplicative) inverse of a matrix
# The @ operator can be used as a shorthand for np.matmul on ndarrays
# .T is the transpose of a matrix

B = np.linalg.inv(X.T @ X) @ (X.T @ Y)
B

array([[ 4.57646455e-02],
       [ 1.88530017e-01],
       [-1.03749304e-03],
       [ 2.93888937e+00]])

In [12]:
# pretty printing
print(f" theta_1= {B[0][0]:.8f},\n theta_2= {B[1][0]:.8f},\n theta_3= {B[2][0]:.8f},\n theta_0= {B[3][0]:.8f}")

 theta_1= 0.04576465,
 theta_2= 0.18853002,
 theta_3= -0.00103749,
 theta_0= 2.93888937


#### 2. Use Linear Regression to solve this problem

In [14]:
X = df[['TV', 'radio', 'newspaper']] # extract relevant features
Y = df[['sales']] # extract sales
model = LinearRegression().fit(X, Y)

In [15]:
# pretty printing
print(f" theta_1= {model.coef_[0][0]:.8f},\n theta_2= {model.coef_[0][1]:.8f},\n theta_3= {model.coef_[0][2]:.8f},\n theta_0= {model.intercept_[0]:.8f}")
r2 = model.score(X, Y)
print("r2: ", r2)

 theta_1= 0.04576465,
 theta_2= 0.18853002,
 theta_3= -0.00103749,
 theta_0= 2.93888937
r2:  0.8972106381789522


#### 3. Compare

Perfect match! Both approaches lead to the same results !!

## Problem 2. Perform the following tasks with Scikit-Learn

#### (1) Scale the features using sklearn.preprocessing.MinMaxScale

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html

In [21]:
data = np.c_[X, Y]

In [22]:
scaler = sk.preprocessing.MinMaxScaler()
print(scaler.fit(data))

MinMaxScaler()


In [23]:
# .data_max_:   Per feature maximum seen in the data
# .data_min_:   Per feature minimum seen in the data
# .data_range_: Per feature range (data_max_ - data_min_) seen in the data

print("max: ", scaler.data_max_)
print("min: ", scaler.data_min_)
print("range: ", scaler.data_range_)

max:  [296.4  49.6 114.   27. ]
min:  [0.7 0.  0.3 1.6]
range:  [295.7  49.6 113.7  25.4]


In [24]:
scaled_data = scaler.transform(data)
scaled_data

array([[0.77578627, 0.76209677, 0.60598065, 0.80708661],
       [0.1481231 , 0.79233871, 0.39401935, 0.34645669],
       [0.0557998 , 0.92540323, 0.60686016, 0.30314961],
       [0.50997633, 0.83266129, 0.51187335, 0.66535433],
       [0.60906324, 0.21774194, 0.51099384, 0.44488189],
       [0.02705445, 0.9858871 , 0.65699208, 0.22047244],
       [0.19208657, 0.66129032, 0.20404573, 0.4015748 ],
       [0.4041258 , 0.39516129, 0.09938434, 0.45669291],
       [0.02671627, 0.04233871, 0.00615655, 0.12598425],
       [0.67331755, 0.05241935, 0.18381706, 0.35433071],
       [0.2211701 , 0.11693548, 0.21020229, 0.27559055],
       [0.72370646, 0.48387097, 0.03254178, 0.62204724],
       [0.07811972, 0.70766129, 0.5769569 , 0.2992126 ],
       [0.32735881, 0.15322581, 0.06068602, 0.31889764],
       [0.68785932, 0.66330645, 0.40193492, 0.68503937],
       [0.65843761, 0.96169355, 0.46262093, 0.81889764],
       [0.22691917, 0.73790323, 1.        , 0.42913386],
       [0.94927291, 0.7983871 ,

#### (2) Split the above scaled data into 80% for training and 20% for test using train_test_split(). Use the same test and training data in the following tasks

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

In [27]:
X = scaled_data[:,:3]

In [28]:
y = scaled_data[:,-1]

In [29]:
#  shuffle=True, random_state=44 : to make the results reproducible 
X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.20, train_size=.80, shuffle=True, random_state=44)

#### (3) use the ordinary linear regression (OLR) to train the model. Compare the R2 scores from training and test data, and discuss your observations

In [31]:
# sk.linearmodel.LinearRegression uses OLR by default (OLR has an aplha of 0 basically) 
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [32]:
print(f"training's data R2: {model.score(X_train, y_train):.5f}")
print(f"testing's data R2: {model.score(X_test, y_test):.5f}")

training's data R2: 0.90659
testing's data R2: 0.83103


Meaning that the model explains approximately 90.65% of the training's data but only 83.1% of the testing dataset

In [34]:
# **** Alternative EXTRA approach to gather metrics ****

y_pred = model.predict(X_test)

mae = sk.metrics.mean_absolute_error(y_test, y_pred)
mse = sk.metrics.mean_squared_error(y_test, y_pred)
r2 = sk.metrics.r2_score(y_test, y_pred)

print(f"Mean Absolute Error: {mae}")
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")

Mean Absolute Error: 0.05171885100739434
Mean Squared Error: 0.004652908399823869
R-squared: 0.8310290458396921


#### (4) Use the Ridge regression to train the model (you may explore different hyperparameter α). Compare the R2 scores from training and test data, and discuss your observation

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html

In [37]:
model = sk.linear_model.Ridge(alpha=.08)
model.fit(X_train, y_train)

print(f"training's data R2: {model.score(X_train, y_train):.5f}")
print(f"testing's data R2: {model.score(X_test, y_test):.5f}")

training's data R2: 0.90656
testing's data R2: 0.83128


With alpha=.08, it seems that the model can explain 90.65% of the training data but only 83.12% of the testing data

#### (5) use the Lasso regression to train the model (you may explore different hyperparameter α). Compare the R2 scores from training and test data, and discuss your observations.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html

In [41]:
model = sk.linear_model.Lasso(alpha=.0002)
model.fit(X_train, y_train)

print(f"training's data R2: {model.score(X_train, y_train):.5f}")
print(f"testing's data R2: {model.score(X_test, y_test):.5f}")

training's data R2: 0.90653
testing's data R2: 0.83132


Meaning that the model explains approximately 90.65% of the training data and 83.13% of the testing data.

It also seems to require a really small alpha so that it is less punishing